core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -404,4 +411,64 @@ private void updateBlockInStream(long blockId) throws IOException {
       mShouldCacheCurrentBlock = mAlluxioStorageType.isStore();
     }
   }
+
+  /**
+   * Seeks to a file position. Blocks are not cached unless they are fully read. This is only called
+   * by {@link FileInStream#seek}.
+   *
+   * @param pos The position to seek to. It is guaranteed to be valid (pos >= 0 && pos != mPos &&
+   *            pos <= mFileLength)
+   * @throws IOException if the seek fails due to an error accessing the stream at the position
+   */
+  private void seekWithoutCachingIncompleteBlocks(long pos) throws IOException {
+    seekBlockInStream(pos);
+    checkAndAdvanceBlockInStream();
+    mCurrentBlockInStream.seek(mPos % mBlockSize);
+  }
+
+  /**
+   * Seeks to a file position. Blocks are cached even if they are fully read. This is only called by
+   * {@link FileInStream#seek}.
+   *
+   * @param pos The position to seek to. It is guaranteed to be valid (pos >= 0 && pos != mPos &&
+   *            pos <= mFileLength).
+   * @throws IOException if the seek fails due to an error accessing the stream at the position
+   */
+  private void seekWithCachingIncompleteBlocks(long pos) throws IOException {
+    boolean isInCurrentBlock = pos / mBlockSize == mPos / mBlockSize;
+    if (mShouldCacheCurrentBlock) {
```

Comment:
I feel this method needs some work to be more clear and easy to reason.

it looks to me several dimensions to make it complicated with multiple cases:
(1) whether this seek is forward or backward
(2) whether the seek is in the same block or not
(3) `mShouldCacheCurrentBlock` is true or not.

Could we have a cleaner branching here so we don't mistakenly get some corner case wrong?